internal/component/database_observability/mysql/collector/query_sample_test.go

```diff
@@ -19,52 +20,318 @@ import (
-	}, 5*time.Second, 100*time.Millisecond)
-
-	collector.Stop()
-	lokiClient.Stop()
-
-	lokiEntries := lokiClient.Received()
-	for _, entry := range lokiEntries {
-		require.Equal(t, model.LabelSet{"job": database_observability.JobName}, entry.Labels)
+			lokiClient := loki_fake.NewClient(func() {})
+
+			collector, err := NewQuerySample(QuerySampleArguments{
+				DB:              db,
+				CollectInterval: time.Minute,
+				EntryHandler:    lokiClient,
+				Logger:          log.NewLogfmtLogger(os.Stderr),
+			})
+			require.NoError(t, err)
+			require.NotNil(t, collector)
+
+			mock.ExpectQuery(selectQuerySamples).WithoutArgs().RowsWillBeClosed().
+				WillReturnRows(
+					sqlmock.NewRows([]string{
+						"digest",
+						"query_sample_text",
+						"query_sample_seen",
+						"query_sample_timer_wait",
+					}).AddRows(
+						tc.rows...,
+					),
+				)
+
+			err = collector.Start(context.Background())
+			require.NoError(t, err)
+
+			require.Eventually(t, func() bool {
+				return len(lokiClient.Received()) == len(tc.logs)
+			}, 5*time.Second, 100*time.Millisecond)
+
+			collector.Stop()
+			lokiClient.Stop()
+
+			lokiEntries := lokiClient.Received()
+			for i, entry := range lokiEntries {
+				require.Equal(t, model.LabelSet{"job": database_observability.JobName}, entry.Labels)
+				require.Equal(t, tc.logs[i], entry.Line)
+			}
+
+			err = mock.ExpectationsWereMet()
+			require.NoError(t, err)
+		})
 	}
-	require.Equal(t, `level=info msg="query samples fetched" op="query_sample" digest="abc123" query_sample_seen="2024-01-01T00:00:00.000Z" query_sample_timer_wait="1000" query_sample_redacted="select * from some_table where id = :redacted1"`, lokiEntries[0].Line)
-	require.Equal(t, `level=info msg="table name parsed" op="query_parsed_table_name" digest="abc123" table="some_table"`, lokiEntries[1].Line)
+}
+
+func TestQuerySampleSqlDriverErrors(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	t.Run("QueryContext() fail", func(t *testing.T) {
+		db, mock, err := sqlmock.New(sqlmock.QueryMatcherOption(sqlmock.QueryMatcherEqual))
+		require.NoError(t, err)
+		defer db.Close()
+
+		lokiClient := loki_fake.NewClient(func() {})
+
+		collector, err := NewQuerySample(QuerySampleArguments{
+			DB:              db,
+			CollectInterval: time.Minute,
+			EntryHandler:    lokiClient,
+			Logger:          log.NewLogfmtLogger(os.Stderr),
+		})
+		require.NoError(t, err)
+		require.NotNil(t, collector)
+
+		mock.ExpectQuery(selectQuerySamples).WithoutArgs().WillReturnError(driver.ErrBadConn)
+
+		err = collector.Start(context.Background())
+		require.NoError(t, err)
 
-	err = mock.ExpectationsWereMet()
-	require.NoError(t, err)
+		require.Eventually(t, func() bool {
+			return collector.Stopped()
+		}, 5*time.Second, 100*time.Millisecond)
+
+		collector.Stop()
+		lokiClient.Stop()
+
+		require.Equal(t, 0, len(lokiClient.Received()))
+
+		err = mock.ExpectationsWereMet()
+		require.NoError(t, err)
+	})
+
+	t.Run("Scan() fail", func(t *testing.T) {
```

Comment:
Can we be more descriptive to the failure mode here, rather than the function call itself which fails